subprojects/language-groovy/src/main/java/org/gradle/api/tasks/compile/GroovyCompile.java

```diff
@@ -60,24 +62,69 @@
 public class GroovyCompile extends AbstractCompile {
     private Compiler<GroovyJavaJointCompileSpec> compiler;
     private FileCollection groovyClasspath;
+    private FileCollection compilerPluginClasspath;
     private final CompileOptions compileOptions;
     private final GroovyCompileOptions groovyCompileOptions = new GroovyCompileOptions();
 
     public GroovyCompile() {
         CompileOptions compileOptions = getServices().get(ObjectFactory.class).newInstance(CompileOptions.class);
         this.compileOptions = compileOptions;
+        this.compilerPluginClasspath = getServices().get(ObjectFactory.class).fileCollection();
         CompilerForkUtils.doNotCacheIfForkingViaExecutable(compileOptions, getOutputs());
     }
 
+    @Override
+    @CompileClasspath
+    public FileCollection getClasspath() {
+        return super.getClasspath();
+    }
+
+    /**
+     * The classpath for compiler plugin to run, for example, AST transformation.
+     *
+     * @since 5.6
+     */
+    @Classpath
+    @Incubating
+    public FileCollection getCompilerPluginClasspath() {
+        return compilerPluginClasspath;
+    }
+
+    /**
+     * The classpath for compiler plugin to run, for example, AST transformation.
+     *
+     * @since 5.6
+     */
+    @Incubating
+    public void setCompilerPluginClasspath(FileCollection compilerPluginClasspath) {
+        if (!experimentalCompilationAvoidanceEnabled()) {
+            throw new IllegalStateException("You need to enable org.gradle.groovy.compilation.avoidance system property!");
+        }
+        this.compilerPluginClasspath = compilerPluginClasspath;
+    }
+
+    private boolean experimentalCompilationAvoidanceEnabled() {
```

Comment:
:o: Why don't we imply that if you add something to `compilerPluginClasspath`, then we enable compile avoidance. In this case instead of failing, we can just warn that this is an experimental feature with known limitations.

Reply:
Added a incubating warning.

Reply:
This is not what I meant. I meant that we can get rid of the system property altogether. If you declare something in the compiler plugins classpath, then it _implies_ that you use the experimental feature, and we can warn.

Reply:
No I think the property is necessary. Imagine you don't want to put anything in the `compilePluginClasspath` (because groovy runtime jar is already `@Classpath`), but you still want to enable this incubating feature, how do you achieve this goal without a property?

Reply:
Ok let's step back. Ideally, the user _must not_ set the compiler plugins classpath by hand. Instead, they will have to use the proper configurations, that, I understand, are going to be created in a separate PR.

So, to declare your AST transformations, you would do something like:

```
dependencies {
    astTransformations "my.ast:xform:1.0"
}
```
then we would have the AST xform classpath inherit from `astTransformations`. So far so good. It, as you mentioned, doesn't let us know the difference between "I say nothing" and "I say it's empty", which, arguably, is annoying...

However, we could say that if you want to use compile avoidance, you _must_ declare the groovy runtime in the `astTransformations` configuration, it would avoid the use of a flag. This corresponds to a reality (the compiler provides AST xforms). Later, we could remove this requirement once it's not experimental anymore, and imply that the Groovy runtime _is_ part of the xform path. WDYT?


```

Reply:
I would say a explicit flag is somewhat better than an magic convention "Put the groovy runtime on a configuration named `astTransformations` then Yay you have compilation avoidance" - the flag is not evil. If it's not experimental anymore we can simply ignore this flag.

Also, I can think of another problem: if we implement as you said, will `astTransformations` configuration be transitive?

Image you have a complicated project structure, `A` dependsOn `B` which depends on `C`, if we declare `compile localGroovy()`, the groovy jar will automatically become `compile` dependencies of `A` and `B`. But what if we change C's declaration to `astTransformations localGroovy()`? Will `A`/`B` have `astTrasformation` or `compile`?